README.txt
= ffi-ncurses 

Author: Sean O'Halpin

A wrapper for ncurses 5.x using ruby-ffi (>= 0.2.0). Tested on Mac OS
X 10.4 (Tiger) and Ubuntu 8.04 with ruby 1.8.6.

The API is very much a transliteration of the C API rather than an
attempt to provide an idiomatic Ruby object-oriented API. The intent
is to provide a 'close to the metal' wrapper around the ncurses
library upon which you can build your own abstractions.

This is still very much a work-in-progress, so expect some rough
edges. Having said that, you can do quite a lot with it as it is. The
main things left to be done are tests, access to global variables and
various macros.

Below are some very preliminary notes on usage. See the examples
directory for real working examples.

== Usage

Load the library with:

  require 'ffi-ncurses'

NCurses methods can be called as module methods:

  begin
    stdscr = NCurses.initscr
    NCurses.clear
    NCurses.addstr("Hello world!")
    NCurses.refresh
    NCurses.getch
  ensure
    NCurses.endwin
  end

or as included methods:

  include NCurses
  begin
    stdscr = initscr
    start_color
    curs_set 0
    raw
    cbreak
    noecho
    clear
    move 10, 10
    standout
    addstr("Hi!")
    standend
  ensure
    endwin
  end

== Set up screen

  require 'ffi-ncurses'
  
  NCurses.initscr
  begin	
    ...
  ensure
    NCurses.endwin
  end

== Typical initialization

  stdscr = NCurses.initscr
  NCurses.start_color
  NCurses.curs_set 0
  NCurses.raw
  NCurses.cbreak
  NCurses.noecho
  NCurses.keypad(stdscr, true)

== Colours
  
  start_color
  init_pair(1, NCurses::Colour::BLACK, NCurses::Colour::RED)
  attr_set NCurses::A_NORMAL, 1, nil
  addch(?A)
  addch(?Z | COLOR_PAIR(1))

== Cursor

=== Turn cursor off

  NCurses.curs_set 0

=== Turn cursor on

  NCurses.curs_set 1

== Windows

  require 'ffi-ncurses'
  begin
    win = newwin(6, 12, 15, 15)
    box(win, 0, 0)
    inner_win = newwin(4, 10, 16, 16)
    waddstr(inner_win, (["Hello window!"] * 5).join(' '))
    wrefresh(win)
    wrefresh(inner_win)
    ch = wgetch(inner_win)

  rescue Object => e
    NCurses.endwin
    puts e
  ensure
    NCurses.endwin
  end

== Mouse handling

The ncurses mouse API is defined in a separate file. To include it use:

  require 'ffi-ncurses/mouse'

You need to specify that you want keypad translation with:

  keypad stdscr, NCurses::TRUE

otherwise your program will receive the raw mouse escape codes,
instead of KEY_MOUSE mouse event codes.

Specify which events you want to handle with:

    mousemask(ALL_MOUSE_EVENTS | REPORT_MOUSE_POSITION, nil)

and set up a mouse event structure to receive the returned values:

    mouse_event = NCurses::MEVENT.new

Receiving mouse events is a two-stage process: first, you are notified
that a mouse event has taken place through a special key code, then
you retrieve the event using +getmouse+. For example:

      ch = getch
      case ch
      when NCurses::KEY_MOUSE
        if getmouse(mouse_event) == NCurses::OK

The mouse event contains the button state (+bstate+) and x, y
coordinates. You can test for the button state using:

  if mouse_event[:bstate] & NCurses::BUTTON1_PRESSED

or

  if NCurses.BUTTON_PRESS(mouse_event[:bstate], 1)

The possible button states are: PRESS, RELEASE, CLICK, DOUBLE_CLICK
and TRIPLE_CLICK.

= Experimental stuff

== Specifying which curses library to use

You can specify which variant of curses you want to use by setting the
environment variable +RUBY_FFI_NCURSES_LIB+ to the one you want. For
example, to use PDCurses X11 curses lib, use:

  RUBY_FFI_NCURSES_LIB=XCurses ruby examples/hello.rb 

You can use this to specify +ncursesw+ for example. Please note that
only the bog standard ncurses lib has been in any way tested as of
yet.

= TO DO

== Turn into a gem

Time to publish as a gem. Need to move files around and make sure it
works as a gem.

== Complete translation of core functions to Darwin (Mac OS X)

There are many macros in darwin ncurses.h which I haven't
implemented yet.

== Work out how to get hold of curscr global variable

There's no way (so far) to access (per process) global variables in
DLLs (hence the hack to stash +stdscr+).

== Tests

This is tricky - I'm not sure exactly how to properly test a wrapper
for a library like ncurses. I certainly don't want to test ncurses!
Instead, I want to ensure my wrapper faithfully reproduces the
functionality of the platform's ncurses lib. To that end, I'm
experimenting with a simple DSL to generate both C and Ruby versions
of a test. With that I can generate equivalent programs and compare
the output. However, this is not really ready for prime time yet.

== Tidy up internals and examples

Things got a bit messy as I switched between the Linux and Mac
versions. The examples should be more focussed.

== Scope implementation of Menu and Form interface wrappers

I'm not particularly interested in the ncurses extension libraries for
forms and menus. I would rather spend time implementing similar
functionality on top of a portable text console library. However, in
the interests of completeness, I suppose I ought to at least scope it out.

= Trivia

While researching ncurses on Google, I innocently entered "curses
getsx" as a search term. Not one for "I'm Feeling Lucky".



History.txt
= Changelog for ffi-ncurses

== 0.4.0 / 2011-09-25

* Supports all functions in ncursesw except the =vwprintw/vwscanw=
  family.
* Added =ACS= constants (box characters). See =examples/acs.rb= and
  =examples/wacs.rb=.
* Added support for =libpanelw=. See =examples/panel_simple.rb=.
* Better examples. See =examples/viewer.rb= for a simple but complete
  file viewing application.
* Methods with boolean arguments now =true= or =false= as well as 1 or
  0.
* Start of a compatibility layer for the existing C extension-based
  Ncurses Ruby libraries - require 'ffi-ncurses/ncurses'.  Runs all
  the examples from ncurses-ruby without changes (except those relying
  on menus and forms). See =examples/ncurses/*.rb=.
* Sets the locale (using FFI::Locale.setlocale(LC_ALL, "")) in ruby
  1.8.x to enable UTF-8 input. See =examples/getkey.rb= to see how to
  distinguish between function keys and Unicode characters.

== 0.3.4 / 2010-08-28

* Added MIT licence (License.txt, COPYING)
* Bumped version
* Added nested PDat struct to WinStruct
* Reformatted comments for rocco

== 0.3.3 / 2010-08-24

* Depends on ffi again (>= 0.6.3)
* Now compatible with ffi-0.6.3
  - thanks and apologies to all those who requested this change months
    ago
* Included keydefs.rb by default
* Examples have been tested with:
  - jruby-1.5.1
  - ree-1.8.7-2010.02
  - ruby-1.8.6-p399
    Note: the examples require "ffi-ncurses/ord-shim" to add
    Integer#ord
  - ruby-1.8.7-p302
  - ruby-1.9.2-p0
* Sadly not compatible with Rubinius (rbx-1.0.0-20100514)
* Removed attempt to load XCurses (PDCurses)
* Removed dependency on bones
  - no Rakefile any more - just use
      gem build ffi-ncurses.gemspec

== 0.3.2 / 2009-02-16

* Bug fix:
  * Reversed regression in 0.3.1 which broke getmaxx, etc. for Mac

== 0.3.1 / 2009-02-16

* Bug fix:
  * Removed dependency on ffi (so gem will install in JRuby)

== 0.3.0 / 2009-02-15

* Features
  * Use FFI::NCurses rather than NCurses
  * Removed dependency on rubygems
  * Reorganised library layout

== 0.2.0 / 2009-02-03

* Features
  * Initial gem release
  * Most features of ncurses wrapped for Linux, Mac, ruby 1.8.6
    and JRuby 1.1.6

* Bugs
  * +newscr+ and +curscr+ cannot be implemented in JRuby until
    find_sym implemented in ffi (expected in JRuby 1.1.7)

== 0.1.0 / 2009-01-19

* Features
  * Initial release


ffi-ncurses.rb
# ruby-ffi wrapper for ncurses
# Sean O'Halpin
# version 0.1.0 - 2008-12-04
# version 0.2.0 - 2009-01-18
# version 0.3.0 - 2009-01-31
# - added stdscr, newscr, curscr
# requires ruby-ffi >= 0.2.0 or jruby >= 1.1.6
# tested with ruby 1.8.6, jruby 1.1.6 on Ubuntu 8.04, Mac OS X 10.4
# ncurses 5.x

require 'ffi'

# would like to check FFI version here rather than use gem but there
# doesn't appear to be a version specified in the ffi lib

module FFI
  module NCurses
    extend FFI::Library

    # use RUBY_FFI_NCURSES_LIB to specify exactly which lib you want, e.g. ncursesw, XCurses (from PDCurses)
    if ENV["RUBY_FFI_NCURSES_LIB"].to_s != ""
      LIB_HANDLE = ffi_lib( ENV["RUBY_FFI_NCURSES_LIB"] ).first
    else
      LIB_HANDLE = ffi_lib( 'ncurses', 'libncurses.so.5', 'XCurses' ).first
    end

    begin
            
      if LIB_HANDLE.respond_to?(:find_symbol)
        # these symbols are defined in ncurses.h
        symbols = [ "stdscr", "curscr", "newscr" ]
        SYMBOLS = Hash[*symbols.map { |sym| [sym, LIB_HANDLE.find_symbol(sym)]}.compact.flatten ]
        SYMBOLS.keys.each do |sym|
          # note that the symbol table entry in a dll for an exported
          # variable contains a ~pointer to the address~ of the variable
          # which will be initialized in the process's bss (uninitialized)
          # data segment when the process is initialized

          # this is unlike methods, where the symbol table entry points to
          # the entry point of the method itself

          # variables need another level of indirection because they are
          # ~not~ shared between process instances - the sharing is of
          # code only
          # p SYMBOLS[sym]
          define_method sym do
            SYMBOLS[sym].read_pointer
          end
          module_function sym
        end
      else
      end
    rescue Object => e
    end
    
    # this list of function signatures was generated by the file
    # generate-ffi-ncurses-function-signatures.rb and inserted here by hand
    functions =
      [
       [:COLOR_PAIR, [:int], :int],
       [:PAIR_NUMBER, [:int], :int],
       # these functions are only in the debug version of ncurses
       #       [:_nc_tracebits, [], :string],
       #       [:_nc_visbuf, [:string], :string],
       #       [:_traceattr2, [:int, :uint], :string],
       #       [:_traceattr, [:uint], :string],
       #       [:_tracecchar_t2, [:int, :pointer], :string],
       #       [:_tracecchar_t, [:pointer], :string],
       #       [:_tracechar, [:int], :string],
       #       [:_tracechtype2, [:int, :uint], :string],
       #       [:_tracechtype, [:uint], :string],
       #       [:_tracedump, [:string, :pointer], :void],
       #       [:_tracef, [:string, :varargs], :void],
       #       [:_tracemouse, [:pointer], :string],
       [:add_wchnstr, [:pointer, :int], :int],
       [:add_wch, [:pointer], :int],
       [:add_wchstr, [:pointer], :int],
       [:addchnstr, [:pointer, :int], :int],
       [:addchstr, [:pointer], :int],
       [:addch, [:uint], :int],
       [:addnstr, [:string, :int], :int],
       [:addstr, [:string], :int],
       [:assume_default_colors, [:int, :int], :int],
       [:attr_get, [:pointer, :pointer, :pointer], :int],
       [:attr_off, [:uint, :pointer], :int],
       [:attr_on, [:uint, :pointer], :int],
       [:attr_set, [:uint, :short, :pointer], :int],
       [:attroff, [:uint], :int],
       [:attron, [:uint], :int],
       [:attrset, [:uint], :int],
       [:baudrate, [], :int],
       [:beep, [], :int],
       [:bkgdset, [:uint], :void],
       [:bkgd, [:uint], :int],
       [:bkgrnd, [:pointer], :int],
       [:bkgrndset, [:pointer], :void],
       [:border_set, [:pointer, :pointer, :pointer, :pointer, :pointer, :pointer, :pointer, :pointer], :int],
       [:border, [:uint, :uint, :uint, :uint, :uint, :uint, :uint, :uint], :int],
       [:box_set, [:pointer, :pointer, :pointer], :int],
       [:box, [:pointer, :uint, :uint], :int],
       [:can_change_color, [], :int],
       [:cbreak, [], :int],
       [:chgat, [:int, :uint, :short, :pointer], :int],
       [:clear, [], :int],
       [:clearok, [:pointer, :int], :int],
       [:clrtobot, [], :int],
       [:clrtoeol, [], :int],
       [:color_content, [:short, :pointer, :pointer, :pointer], :int],
       [:color_set, [:short, :pointer], :int],
       [:copywin, [:pointer, :pointer, :int, :int, :int, :int, :int, :int, :int], :int],
       [:curs_set, [:int], :int],
       [:curses_version, [], :string],
       [:def_prog_mode, [], :int],
       [:def_shell_mode, [], :int],
       [:define_key, [:string, :int], :int],
       [:delay_output, [:int], :int],
       [:delch, [], :int],
       [:deleteln, [], :int],
       [:delscreen, [:pointer], :void],
       [:delwin, [:pointer], :int],
       [:derwin, [:pointer, :int, :int, :int, :int], :pointer],
       [:doupdate, [], :int],
       [:dupwin, [:pointer], :pointer],
       [:echo_wchar, [:pointer], :int],
       [:echochar, [:uint], :int],
       [:echo, [], :int],
       [:endwin, [], :int],
       [:erasechar, [], :char],
       [:erase, [], :int],
       [:filter, [], :void],
       [:flash, [], :int],
       [:flushinp, [], :int],
       [:getbkgd, [:pointer], :uint],
       [:getbkgrnd, [:pointer], :int],
       [:getch, [], :int],

       [:getattrs, [:pointer], :int],
       [:getcurx, [:pointer], :int],
       [:getcury, [:pointer], :int],
       [:getbegx, [:pointer], :int],
       [:getbegy, [:pointer], :int],
       [:getmaxx, [:pointer], :int],
       [:getmaxy, [:pointer], :int],
       [:getparx, [:pointer], :int],
       [:getpary, [:pointer], :int],

       # this is defined in mouse.rb
       #      [:getmouse, [:pointer], :int],
       [:getnstr, [:string, :int], :int],
       [:getstr, [:string], :int],
       [:getwin, [:pointer], :pointer],
       [:halfdelay, [:int], :int],
       [:has_colors, [], :int],
       [:has_ic, [], :int],
       [:has_il, [], :int],
       [:has_key, [:int], :int],
       [:hline_set, [:pointer, :int], :int],
       [:hline, [:uint, :int], :int],
       [:idcok, [:pointer, :int], :void],
       [:idlok, [:pointer, :int], :int],
       [:immedok, [:pointer, :int], :void],
       [:in_wchnstr, [:pointer, :int], :int],
       [:in_wch, [:pointer], :int],
       [:in_wchstr, [:pointer], :int],
       [:inchnstr, [:pointer, :int], :int],
       [:inchstr, [:pointer], :int],
       [:inch, [], :uint],
       [:init_color, [:short, :short, :short, :short], :int],
       [:init_pair, [:short, :short, :short], :int],
       # need to intercept this to get stdscr for JRuby (for the moment)
       ["_initscr", :initscr, [], :pointer],
       [:innstr, [:string, :int], :int],
       [:ins_wch, [:pointer], :int],
       [:insch, [:uint], :int],
       [:insdelln, [:int], :int],
       [:insertln, [], :int],
       [:insnstr, [:string, :int], :int],
       [:insstr, [:string], :int],
       [:instr, [:string], :int],
       [:intrflush, [:pointer, :int], :int],
       [:is_linetouched, [:pointer, :int], :int],
       [:is_term_resized, [:int, :int], :int],
       [:is_wintouched, [:pointer], :int],
       [:isendwin, [], :int],
       [:key_defined, [:string], :int],
       [:keybound, [:int, :int], :string],
       [:keyname, [:int], :string],
       [:keyok, [:int, :int], :int],
       [:keypad, [:pointer, :int], :int],
       [:killchar, [], :char],
       [:leaveok, [:pointer, :int], :int],
       [:longname, [], :string],
       [:mcprint, [:string, :int], :int],
       [:meta, [:pointer, :int], :int],
       [:mouse_trafo, [:pointer, :pointer, :int], :int],
       [:mouseinterval, [:int], :int],
       [:mousemask, [:uint, :pointer], :uint],
       [:move, [:int, :int], :int],
       [:mvadd_wch, [:int, :int, :pointer], :int],
       [:mvadd_wchnstr, [:int, :int, :pointer, :int], :int],
       [:mvadd_wchstr, [:int, :int, :pointer], :int],
       [:mvaddch, [:int, :int, :uint], :int],
       [:mvaddchnstr, [:int, :int, :pointer, :int], :int],
       [:mvaddchstr, [:int, :int, :pointer], :int],
       [:mvaddnstr, [:int, :int, :string, :int], :int],
       [:mvaddstr, [:int, :int, :string], :int],
       [:mvchgat, [:int, :int, :int, :uint, :short, :pointer], :int],
       [:mvcur, [:int, :int, :int, :int], :int],
       [:mvdelch, [:int, :int], :int],
       [:mvderwin, [:pointer, :int, :int], :int],
       [:mvgetch, [:int, :int], :int],
       [:mvgetnstr, [:int, :int, :string, :int], :int],
       [:mvgetstr, [:int, :int, :string], :int],
       [:mvhline_set, [:int, :int, :pointer, :int], :int],
       [:mvhline, [:int, :int, :uint, :int], :int],
       [:mvin_wch, [:int, :int, :pointer], :int],
       [:mvin_wchnstr, [:int, :int, :pointer, :int], :int],
       [:mvin_wchstr, [:int, :int, :pointer], :int],
       [:mvinch, [:int, :int], :uint],
       [:mvinchnstr, [:int, :int, :pointer, :int], :int],
       [:mvinchstr, [:int, :int, :pointer], :int],
       [:mvinnstr, [:int, :int, :string, :int], :int],
       [:mvins_wch, [:int, :int, :pointer], :int],
       [:mvinsch, [:int, :int, :uint], :int],
       [:mvinsnstr, [:int, :int, :string, :int], :int],
       [:mvinsstr, [:int, :int, :string], :int],
       [:mvinstr, [:int, :int, :string], :int],
       [:mvprintw, [:int, :int, :string, :varargs], :int],
       [:mvscanw, [:int, :int, :string, :varargs], :int],
       [:mvvline_set, [:int, :int, :pointer, :int], :int],
       [:mvvline, [:int, :int, :uint, :int], :int],
       [:mvwadd_wchnstr, [:pointer, :int, :int, :pointer, :int], :int],
       [:mvwadd_wch, [:pointer, :int, :int, :pointer], :int],
       [:mvwadd_wchstr, [:pointer, :int, :int, :pointer], :int],
       [:mvwaddchnstr, [:pointer, :int, :int, :pointer, :int], :int],
       [:mvwaddch, [:pointer, :int, :int, :uint], :int],
       [:mvwaddchstr, [:pointer, :int, :int, :pointer], :int],
       [:mvwaddnstr, [:pointer, :int, :int, :string, :int], :int],
       [:mvwaddstr, [:pointer, :int, :int, :string], :int],
       [:mvwchgat, [:pointer, :int, :int, :int, :uint, :short, :pointer], :int],
       [:mvwdelch, [:pointer, :int, :int], :int],
       [:mvwgetch, [:pointer, :int, :int], :int],
       [:mvwgetnstr, [:pointer, :int, :int, :string, :int], :int],
       [:mvwgetstr, [:pointer, :int, :int, :string], :int],
       [:mvwhline_set, [:pointer, :int, :int, :pointer, :int], :int],
       [:mvwhline, [:pointer, :int, :int, :uint, :int], :int],
       [:mvwin_wchnstr, [:pointer, :int, :int, :pointer, :int], :int],
       [:mvwin_wch, [:pointer, :int, :int, :pointer], :int],
       [:mvwin_wchstr, [:pointer, :int, :int, :pointer], :int],
       [:mvwinchnstr, [:pointer, :int, :int, :pointer, :int], :int],
       [:mvwinch, [:pointer, :int, :int], :uint],
       [:mvwinchstr, [:pointer, :int, :int, :pointer], :int],
       [:mvwinnstr, [:pointer, :int, :int, :string, :int], :int],
       [:mvwin, [:pointer, :int, :int], :int],
       [:mvwins_wch, [:pointer, :int, :int, :pointer], :int],
       [:mvwinsch, [:pointer, :int, :int, :uint], :int],
       [:mvwinsnstr, [:pointer, :int, :int, :string, :int], :int],
       [:mvwinsstr, [:pointer, :int, :int, :string], :int],
       [:mvwinstr, [:pointer, :int, :int, :string], :int],
       [:mvwprintw, [:pointer, :int, :int, :string, :varargs], :int],
       [:mvwscanw, [:pointer, :int, :int, :string, :varargs], :int],
       [:mvwvline_set, [:pointer, :int, :int, :pointer, :int], :int],
       [:mvwvline, [:pointer, :int, :int, :uint, :int], :int],
       [:napms, [:int], :int],
       [:newpad, [:int, :int], :pointer],
       [:newterm, [:string, :pointer, :pointer], :pointer],
       [:newwin, [:int, :int, :int, :int], :pointer],
       [:nl, [], :int],
       [:nocbreak, [], :int],
       [:nodelay, [:pointer, :int], :int],
       [:noecho, [], :int],
       [:nonl, [], :int],
       [:noqiflush, [], :void],
       [:noraw, [], :int],
       [:notimeout, [:pointer, :int], :int],
       [:overlay, [:pointer, :pointer], :int],
       [:overwrite, [:pointer, :pointer], :int],
       [:pair_content, [:short, :pointer, :pointer], :int],
       [:pecho_wchar, [:pointer, :pointer], :int],
       [:pechochar, [:pointer, :uint], :int],
       [:pnoutrefresh, [:pointer, :int, :int, :int, :int, :int, :int], :int],
       [:prefresh, [:pointer, :int, :int, :int, :int, :int, :int], :int],
       [:printw, [:string, :varargs], :int],
       [:putp, [:string], :int],
       [:putwin, [:pointer, :pointer], :int],
       [:qiflush, [], :void],
       [:raw, [], :int],
       [:redrawwin, [:pointer], :int],
       [:refresh, [], :int],
       [:reset_prog_mode, [], :int],
       [:reset_shell_mode, [], :int],
       [:resetty, [], :int],
       [:resize_term, [:int, :int], :int],
       [:resizeterm, [:int, :int], :int],
       [:ripoffline, [:int, :pointer], :int],
       [:savetty, [], :int],
       [:scanw, [:string, :varargs], :int],
       [:scr_dump, [:string], :int],
       [:scr_init, [:string], :int],
       [:scr_restore, [:string], :int],
       [:scr_set, [:string], :int],
       [:scrl, [:int], :int],
       [:scrollok, [:pointer, :int], :int],
       [:scroll, [:pointer], :int],
       [:set_term, [:pointer], :pointer],
       [:setscrreg, [:int, :int], :int],
       [:slk_attr_off, [:uint, :pointer], :int],
       [:slk_attr_on, [:uint, :pointer], :int],
       [:slk_attr_set, [:uint, :short, :pointer], :int],
       [:slk_attroff, [:uint], :int],
       [:slk_attron, [:uint], :int],
       [:slk_attrset, [:uint], :int],
       [:slk_attr, [], :uint],
       [:slk_clear, [], :int],
       [:slk_color, [:short], :int],
       [:slk_init, [:int], :int],
       [:slk_label, [:int], :string],
       [:slk_noutrefresh, [], :int],
       [:slk_refresh, [], :int],
       [:slk_restore, [], :int],
       [:slk_set, [:int, :string, :int], :int],
       [:slk_touch, [], :int],
       [:standend, [], :int],
       [:standout, [], :int],
       [:start_color, [], :int],
       [:subpad, [:pointer, :int, :int, :int, :int], :pointer],
       [:subwin, [:pointer, :int, :int, :int, :int], :pointer],
       [:syncok, [:pointer, :int], :int],
       [:term_attrs, [], :uint],
       [:termattrs, [], :uint],
       [:termname, [], :string],
       [:tigetflag, [:string], :int],
       [:tigetnum, [:string], :int],
       [:tigetstr, [:string], :string],
       [:timeout, [:int], :void],
       [:touchline, [:pointer, :int, :int], :int],
       [:touchwin, [:pointer], :int],
       [:tparm, [:string, :varargs], :string],
       #       [:trace, [:uint], :void],
       [:typeahead, [:int], :int],
       [:ungetch, [:int], :int],
       #       [:ungetmouse, [:pointer], :int],
       [:untouchwin, [:pointer], :int],
       [:use_default_colors, [], :int],
       [:use_env, [:int], :void],
       [:use_extended_names, [:int], :int],
       [:vid_attr, [:uint, :short, :pointer], :int],
       [:vid_puts, [:uint, :short, :pointer, :pointer], :int],
       [:vidattr, [:uint], :int],
       [:vidputs, [:uint, :pointer], :int],
       [:vline_set, [:pointer, :int], :int],
       [:vline, [:uint, :int], :int],
       [:wadd_wchnstr, [:pointer, :pointer, :int], :int],
       [:wadd_wch, [:pointer, :pointer], :int],
       [:wadd_wchstr, [:pointer, :pointer], :int],
       [:waddchnstr, [:pointer, :pointer, :int], :int],
       [:waddch, [:pointer, :uint], :int],
       [:waddchstr, [:pointer, :pointer], :int],
       [:waddnstr, [:pointer, :string, :int], :int],
       [:waddstr, [:pointer, :string], :int],
       [:wattr_get, [:pointer, :pointer, :pointer, :pointer], :int],
       [:wattr_off, [:pointer, :uint, :pointer], :int],
       [:wattr_on, [:pointer, :uint, :pointer], :int],
       [:wattr_set, [:pointer, :uint, :short, :pointer], :int],
       [:wattroff, [:pointer, :int], :int],
       [:wattron, [:pointer, :int], :int],
       [:wattrset, [:pointer, :int], :int],
       [:wbkgd, [:pointer, :uint], :int],
       [:wbkgdset, [:pointer, :uint], :void],
       [:wbkgrnd, [:pointer, :pointer], :int],
       [:wbkgrndset, [:pointer, :pointer], :void],
       [:wborder_set, [:pointer, :pointer, :pointer, :pointer, :pointer, :pointer, :pointer, :pointer, :pointer], :int],
       [:wborder, [:pointer, :uint, :uint, :uint, :uint, :uint, :uint, :uint, :uint], :int],
       [:wchgat, [:pointer, :int, :uint, :short, :pointer], :int],
       [:wclear, [:pointer], :int],
       [:wclrtobot, [:pointer], :int],
       [:wclrtoeol, [:pointer], :int],
       [:wcolor_set, [:pointer, :short, :pointer], :int],
       [:wcursyncup, [:pointer], :void],
       [:wdelch, [:pointer], :int],
       [:wdeleteln, [:pointer], :int],
       [:wecho_wchar, [:pointer, :pointer], :int],
       [:wechochar, [:pointer, :uint], :int],
       [:wenclose, [:pointer, :int, :int], :int],
       [:werase, [:pointer], :int],
       [:wgetbkgrnd, [:pointer, :pointer], :int],
       [:wgetch, [:pointer], :int],
       [:wgetnstr, [:pointer, :string, :int], :int],
       [:wgetstr, [:pointer, :string], :int],
       [:whline_set, [:pointer, :pointer, :int], :int],
       [:whline, [:pointer, :uint, :int], :int],
       [:win_wchnstr, [:pointer, :pointer, :int], :int],
       [:win_wch, [:pointer, :pointer], :int],
       [:win_wchstr, [:pointer, :pointer], :int],
       [:winchnstr, [:pointer, :pointer, :int], :int],
       [:winch, [:pointer], :uint],
       [:winchstr, [:pointer, :pointer], :int],
       [:winnstr, [:pointer, :string, :int], :int],
       [:wins_wch, [:pointer, :pointer], :int],
       [:winsch, [:pointer, :uint], :int],
       [:winsdelln, [:pointer, :int], :int],
       [:winsertln, [:pointer], :int],
       [:winsnstr, [:pointer, :string, :int], :int],
       [:winsstr, [:pointer, :string], :int],
       [:winstr, [:pointer, :string], :int],
       [:wmove, [:pointer, :int, :int], :int],
       [:wnoutrefresh, [:pointer], :int],
       [:wprintw, [:pointer, :string, :varargs], :int],
       [:wredrawln, [:pointer, :int, :int], :int],
       [:wrefresh, [:pointer], :int],
       [:wresize, [:pointer, :int, :int], :int],
       [:wscanw, [:pointer, :string, :varargs], :int],
       [:wscrl, [:pointer, :int], :int],
       [:wsetscrreg, [:pointer, :int, :int], :int],
       [:wstandend, [:pointer], :int],
       [:wstandout, [:pointer], :int],
       [:wsyncdown, [:pointer], :void],
       [:wsyncup, [:pointer], :void],
       [:wtimeout, [:pointer, :int], :void],
       [:wtouchln, [:pointer, :int, :int, :int], :int],
       [:wvline_set, [:pointer, :pointer, :int], :int],
       [:wvline, [:pointer, :uint, :int], :int]
      ]
    # end of autogenerated function list

    @unattached_functions = []
    class << self
      def unattached_functions
        @unattached_functions
      end
    end

    # attach functions
    functions.each do |func|
      begin
        attach_function(*func)
      rescue Object => e
        # for debugging
        unattached_functions << func[0]
      end
    end

    module Colour
      COLOR_BLACK   = BLACK   = 0
      COLOR_RED     = RED     = 1
      COLOR_GREEN   = GREEN   = 2
      COLOR_YELLOW  = YELLOW  = 3
      COLOR_BLUE    = BLUE    = 4
      COLOR_MAGENTA = MAGENTA = 5
      COLOR_CYAN    = CYAN    = 6
      COLOR_WHITE   = WHITE   = 7
    end
    Color = Colour
    include Colour

    # FIXME: remove this code when JRuby gets find_sym
    # fixup for JRuby 1.1.6 - doesn't have find_sym
    # can hack for stdscr but not curscr or newscr (no methods return them)
    # this will all be removed when JRuby 1.1.7 is released
    module FixupInitscr
      if !NCurses.respond_to?(:stdscr)
        def initscr
          @stdscr = NCurses._initscr
        end
        def stdscr
          @stdscr
        end
      else
        def initscr
          NCurses._initscr
        end
      end
    end
    include FixupInitscr
    
    module Attributes
      # following definitions have been copied (almost verbatim) from ncurses.h
      NCURSES_ATTR_SHIFT = 8
      def self.NCURSES_BITS(mask, shift)
        ((mask) << ((shift) + NCURSES_ATTR_SHIFT))
      end

      WA_NORMAL     = A_NORMAL     = (1 - 1)
      WA_ATTRIBUTES = A_ATTRIBUTES = NCURSES_BITS(~(1 - 1),0)
      WA_CHARTEXT   = A_CHARTEXT   = (NCURSES_BITS(1,0) - 1)
      WA_COLOR      = A_COLOR      = NCURSES_BITS(((1) << 8) - 1,0)
      WA_STANDOUT   = A_STANDOUT   = NCURSES_BITS(1,8)  # best highlighting mode available
      WA_UNDERLINE  = A_UNDERLINE  = NCURSES_BITS(1,9)  # underlined text
      WA_REVERSE    = A_REVERSE    = NCURSES_BITS(1,10) # reverse video
      WA_BLINK      = A_BLINK      = NCURSES_BITS(1,11) # blinking text
      WA_DIM        = A_DIM        = NCURSES_BITS(1,12) # half-bright text
      WA_BOLD       = A_BOLD       = NCURSES_BITS(1,13) # extra bright or bold text
      WA_ALTCHARSET = A_ALTCHARSET = NCURSES_BITS(1,14)
      WA_INVIS      = A_INVIS      = NCURSES_BITS(1,15)
      WA_PROTECT    = A_PROTECT    = NCURSES_BITS(1,16)
      WA_HORIZONTAL = A_HORIZONTAL = NCURSES_BITS(1,17)
      WA_LEFT       = A_LEFT       = NCURSES_BITS(1,18)
      WA_LOW        = A_LOW        = NCURSES_BITS(1,19)
      WA_RIGHT      = A_RIGHT      = NCURSES_BITS(1,20)
      WA_TOP        = A_TOP        = NCURSES_BITS(1,21)
      WA_VERTICAL   = A_VERTICAL   = NCURSES_BITS(1,22)
    end
    include Attributes

    module Constants
      FALSE = 0
      TRUE = 1

      ERR = -1
      OK = 0
    end
    include Constants

    module EmulatedFunctions
      
      # these 'functions' are implemented as macros in ncurses
      
      # I'm departing from the NCurses API here - makes no sense to
      # force people to use pointer return values when these are
      # implemented as macros to make them easy to use in C when we have
      # multiple return values in Ruby for that exact same purpose
      def getyx(win)
        [NCurses.getcury(win), NCurses.getcurx(win)]
      end
      def getbegyx(win)
        [NCurses.getbegy(win), NCurses.getbegx(win)]
      end
      def getparyx(win)
        [NCurses.getpary(win), NCurses.getparx(win)]
      end
      def getmaxyx(win)
        [NCurses.getmaxy(win), NCurses.getmaxx(win)]
      end

      # transliterated from ncurses.h
      # note: had to spell out NCurses::TRUE - will look into this
      def getsyx
        if (is_leaveok(newscr) == NCurses::TRUE)
			    [-1, -1]
        else
          getyx(newscr)
        end
      end
      def setsyx(y, x)
        if y == -1 && x == -1
			    leaveok(newscr, NCurses::TRUE)
        else
			    leaveok(newscr, NCurses::FALSE)
			    wmove(newscr, y, x)
        end
      end
      
      def self.fixup(function, &block)
        if NCurses.unattached_functions.include?(function)
          block.call
        end
      end

      # hack for XCurses (PDCurses 3.3) - many more to come I suspect :)
      fixup :getch do
        def getch
          wgetch(stdscr)
        end
      end
    end
    include EmulatedFunctions

    # fixes for Mac OS X (mostly macros directly referencing the WINDOW struct)
    if ::FFI::Platform::OS == "darwin"
      require 'ffi-ncurses-darwin'
      include NCurses::Darwin
    end

    # make all those instance methods module methods too
    extend self

  end
end

